apps/web/src/app/(private)/datasets/generate/GenerateDatasetContent.tsx

```diff
@@ -0,0 +1,312 @@
+        }).then((r) => r.unwrap())
+        setPreviewCsv(parsedCsv)
+      }
+    },
+  )
+
+  const handleProjectChange = (projectId: number) => {
+    setProjectId(projectId)
+  }
+
+  const handleDocumentChange = (newDocumentUuid: string) => {
+    setDocumentUuid(newDocumentUuid)
+  }
+
+  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault()
+    const formData = new FormData(e.target as HTMLFormElement)
+
+    if (!previewCsv) {
+      await runPreviewAction({
+        parameters: formData.get('parameters') as string,
+        description: formData.get('description') as string,
+      })
+    } else {
+      const response = await runGenerateAction({
+        parameters: formData.get('parameters') as string,
+        description: formData.get('description') as string,
+        rowCount: parseInt(formData.get('rows') as string, 10),
+        name: formData.get('name') as string,
+      })
+
+      try {
+        const dataset = await response
+        if (!dataset) return
+
+        mutate([...datasets, dataset])
+        navigate.push(ROUTES.datasets.root)
+      } catch (error) {
+        console.error(error)
+      }
+    }
+  }
+
+  const handleRegeneratePreview = async () => {
+    const form = window.document.getElementById(
+      'generateDatasetForm',
+    ) as HTMLFormElement
+    const formData = new FormData(form)
+
+    await runPreviewAction({
+      parameters: formData.get('parameters') as string,
+      description: formData.get('description') as string,
+    })
+  }
+
+  useEffect(() => {
+    const fetchMetadata = async () => {
+      if (document) {
+        const metadata = await readMetadata({
+          prompt: document.resolvedContent ?? '',
+          fullPath: document.path,
+        })
+
+        setMetadata(metadata)
+      }
+    }
+
+    fetchMetadata()
+  }, [document])
+
+  return (
+    <Modal
+      open
+      size='large'
+      onOpenChange={(open) => !open && navigate.push(ROUTES.datasets.root)}
+      title='Generate new dataset'
+      description='Generate a dataset of parameters using AI from one of your prompts. Datasets can be used to run batch evaluations over prompts.'
+      footer={
+        <>
+          <CloseTrigger />
+          {previewCsv && (
+            <Button
+              onClick={handleRegeneratePreview}
+              disabled={previewIsLoading || generateIsLoading}
+              fancy
+              variant='outline'
+            >
+              Regenerate preview
+            </Button>
+          )}
+          <Button
+            disabled={!documentUuid || previewIsLoading || generateIsLoading}
+            fancy
+            form='generateDatasetForm'
+            type='submit'
+          >
+            {previewIsLoading
+              ? 'Generating...'
+              : generateIsLoading
+                ? 'Generating...'
+                : previewCsv
+                  ? 'Generate dataset'
+                  : 'Generate preview'}
```

Comment:
When already generated, the button should say `Save dataset` instead, because `Generate dataset` sounds like it's just going to reuse the user prompt to generate a new dataset, but instead it closes the modal and saves the dataset.

Reply:
hmm actually it does generate a new dataset! the preview is fast because we ask it to generate only 10 rows, then the generate dataset generates the full csv. Maybe it should say "generate & save data"